custom NER features CRF

jamesmf opened the pull request:
This is a work-in-progress implementation that allows custom entity-featurizers to pass features along to `CRFEntityExtractor`. This would enable new featurizers that passed word/token vectors along to `CRFEntityExtractor`. Also following this `ner_features` pattern would make it easy to develop other components that could also do custom NER. 

The edit to the example is just to show that, while the `DummyNERFeaturizer` provides nothing but noise to the `CRFEntityExtractor`, it seems to be working.

Just looking for feedback before I add tests or do much else.

Addresses https://forum.rasa.com/t/feeding-custom-pretrained-embeddings-for-ner-crf/5406

**Proposed changes**:
- Make `CRFEntityExtractor` look for `ner_features` on the message
- Have `CRFEntityExtractor` convert the features from an array to `python-crfsuite` style dicts

**Status (please check what you already did)**:
- [X] made PR ready for code review
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa_nlu#code-style) for instructions)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4187) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/RasaHQ/rasa?pullRequest=4187) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RasaHQ/rasa?pullRequest=4187) it.</sub>

jamesmf commented:
For this dummy featurizer that simply adds a `10d` random vector to the `ner_features` attribute, this is how it gets featurized for `python-crfsuite`
```
{
  "custom_ner_ft": {
    "0": 0.6106456639385146,
    "1": 0.7908092850036196,
    "2": 0.9450676837200174,
    "3": 0.8118531808398548,
    "4": 0.9868772677831507,
    "5": 0.3286065567479264,
    "6": 0.5570708725716922,
    "7": 0.6820622317927959,
    "8": 0.11855285847390751,
    "9": 0.26320288572723727
  }
}

```

tabergma commented:
@jamesmf Thanks for submitting the PR! I like the idea. I did not had time to take a closer look yet, but will provide some feedback by latest end of the week.

jamesmf commented:
https://github.com/RasaHQ/rasa/issues/4214 
example above where adding this functionality improves NER in `examples/restaurantbot`

Let me know what kinds of tests I should add, and whether we should move the default features out of `CRFEntityExtractor` and into a `CRFEntityFeaturizer` class.

jamesmf commented on rasa/nlu/featurizers/dummy_ner_featurizer.py:
```diff
@@ -0,0 +1,69 @@
+import logging
```
This file can be deleted, was just showing an example of the minimal requirement to create a Featurizer for this

jamesmf commented:
> Great start! I like the idea 👍
> 
> Added a few style comments.
> 
> You removed the custom featurizer and went for taking the token vector from spacy as additional NER features. That is definitely easier for our users, as they just need to add a keyword to the features of the CRF. However, why did you decided to remove the custom featurizer? Do you think this feature is not needed?
> 
> As the `ner_features` are basically word embeddings, we might want to rename `ner_features` to `word_embedding`. What do you think?
> 
> Can you please also add a Changelog entry and update the documentation (https://github.com/RasaHQ/rasa/blob/master/docs/nlu/entity-extraction.rst)? You should also add a test that makes use of the new features. Thanks.

Thank you so much for the full review! I'll make those changes.

The thing I deleted was a "dummy" featurizer just to show that it could be added; it just added a random 10-d vector to each token. I think what I'd want is to update the documentation to explain how to add a custom featurizer. I'm also open to removing the default feature options from the `CRFEntityExtractor` and implement those as a `Featurizer`. But that's a bit of a change, and I don't want to do that unless you think it's the right way to go.

tabergma commented:
If we keep it as it is and I would add a custom featurizer to add my own word embeddings to `ner_features`, we would end up having the spacy word vector and my own word embeddings as features. Correct? That does not seem right. I would rather want to overwrite those, I guess. Maybe, we should separate them: Add the option to put any custom new features to `ner_features`. Plus add the option to enable `spacy_word_embeddings`, which is a different feature. Not sure. What do you think?

jamesmf commented:
Oh! yes, I meant to implement "add_ner_features" on the `SpacyFeaturizer` as optional

My intended functionality was:
- If you use `SpacyFeaturizer`, you get `ner_features` added for free, if you want them (on by default, controlled by a flag on the featurizer)
- If you add a custom featurizer and also have a `SpacyFeaturizer`, the two will combine, just as they would if you added a custom intent featurizer.
- If you don't want that, you simply specify `add_ner_features: false` to the config under `SpacyFeaturizer`.

tabergma commented:
Sounds good. I guess, I just wait until you update the PR. Just ping me once you are done with the changes. Thanks.

tabergma commented:
One other thing: Can you please evaluate the performance improvement on the Sara bot (https://github.com/RasaHQ/rasa-demo)? The dataset is much bigger compared to the restaurant bot you used so far. Would be interesting to know how big the impact actually is. Thanks.

jamesmf commented:
Reran the same experiment (25% and 50% hold-out, 5 runs, one with default features for NER, one with the same features + `spacy` `en_web_core_md` vectors as features too
```
{
  "25%": {
    "micro avg": {
      "config": 0.9791423487292383,
      "config_orig": 0.976507394601747
    },
    "macro avg": {
      "config": 0.6321974646757288,
      "config_orig": 0.557042332107
    },
    "weighted avg": {
      "config": 0.9757829656864989,
      "config_orig": 0.9697939177484041
    }
  },
  "50%": {
    "micro avg": {
      "config": 0.976118992270199,
      "config_orig": 0.9741663333955127
    },
    "macro avg": {
      "config": 0.5345575061878215,
      "config_orig": 0.4813566221323427
    },
    "weighted avg": {
      "config": 0.9705968761047593,
      "config_orig": 0.9659821262255277
    }
  }
}

```
The new feature set wins out almost across the board (slightly worse on `entity` and `product` entities in a some runs), though not by a huge amount. Unsurprisingly, the word-vector version does a much better job with `name` and `company`.

jamesmf commented:
I'm not 100% sure on how strict `provides` is on a `Component`.

I want to make `SpacyFeaturizer` only add its word vectors to the message if you specify that it should in the config, but if I have that optional behavior, is it wrong to have a `provides = ['ner_features']`?  Or is it more correct to leave the `provides` in there, but to add a blank `ner_features` to the `message` if you specify that you don't want the word vectors?

tabergma commented:
I would go with the second option: Add `['ner_features']` to the return list of `provides`. By default `ner_features` is just an empty list. Only if you specify it in the config file, the `ner_features` are actually set.

jamesmf commented:
Awesome, I'll fix that and then turn to the existing tests to get an example of how to test this.

Thanks!

jamesmf commented:
I added the functionality discussed to `SpacyFeaturizer`. This was the first `component_config` entry in `SpacyFeaturizer` which meant adding an `__init__()`, so I'd love a pair of eyes on that code to make sure it makes sense to add `__init__()` but not need custom `persist()` or `load()`.

I also added tests that ensure that when you ask for `ner_features` on the `SpacyFeaturizer`, you get them (followed the example of the test above it for the spacy intent classifier features), and also that when you don't want `ner_features`, it still sets `ner_features` on the message, but as a list of empty lists of `len(tokens)` (which will combine with other `ner_features` just fine using `np.concatenate`).

tabergma commented:
On a first glance it looks fine. Once you are done with all your changes, I'll have a closer look. Just let me know when you are done. Thanks.

jamesmf commented:
I'm done with all my intended changes, I'd love feedback and also suggestions for raising the coverage %

tabergma commented on rasa/nlu/extractors/crf_entity_extractor.py:
```diff
@@ -529,27 +573,64 @@ def __tag_of_token(token):
+    def __additional_ner_features(message: Message) -> List[Any]:
+        features = message.get("ner_features", [])
+        tokens = message.get("tokens", [])
+        # perhaps an assertion here instead? following pattern above.
```
What do you mean by assertion? cannot find anything like that above? We should either decide to implement it or remove the comment.

jamesmf replied:
oh yeah I worded that poorly. I meant "maybe we should have an assertion here, BUT I'm following the pattern above instead". I'll remove the comment if you think it looks okay

tabergma commented on rasa/nlu/extractors/crf_entity_extractor.py:
```diff
@@ -529,27 +573,64 @@ def __tag_of_token(token):
+        tokens = message.get("tokens", [])
+        # perhaps an assertion here instead? following pattern above.
+        if len(tokens) != len(features):
+            warn_string = "Number of tokens for custom NER features ({}) does not match number of tokens ({})".format(
```
```suggestion
            warn_string = "Number of custom NER features ({}) does not match number of tokens ({})".format(
```

tabergma commented on rasa/nlu/extractors/crf_entity_extractor.py:
```diff
@@ -529,27 +573,64 @@ def __tag_of_token(token):
+        features_out = []
+        for feature in features:
+            feature_dict = {str(ind): ft for ind, ft in enumerate(feature)}
+            converted = {"custom_ner_ft": feature_dict}
```
please rename `custom_ner_ft` to `custom_ner_features`

tabergma commented on rasa/nlu/extractors/crf_entity_extractor.py:
```diff
@@ -529,27 +573,64 @@ def __tag_of_token(token):
+        # convert to python-crfsuite feature format
+        features_out = []
+        for feature in features:
+            feature_dict = {str(ind): ft for ind, ft in enumerate(feature)}
```
```suggestion
            feature_dict = {str(index): token_features for index, token_features in enumerate(feature)}
```

tabergma commented on rasa/nlu/extractors/crf_entity_extractor.py:
```diff
@@ -529,27 +573,64 @@ def __tag_of_token(token):
             tokens = message.get("spacy_doc")
         else:
             tokens = message.get("tokens")
+        if self.use_ner_features:
```
```suggestion

        ner_features = self.__additional_ner_features(message) if self.use_ner_features else None
```

tabergma commented on rasa/nlu/extractors/crf_entity_extractor.py:
```diff
@@ -529,27 +573,64 @@ def __tag_of_token(token):
         else:
             tokens = message.get("tokens")
+        if self.use_ner_features:
+            ner_features = self.__additional_ner_features(message)
```
```suggestion

```

tabergma commented:
I guess, you can also make this PR a "real" PR. It is not really a draft anymore :)

tabergma commented on docs/nlu/components.rst:
```diff
@@ -116,6 +116,15 @@ SpacyFeaturizer
+        pipeline:
+        - name: "SpacyFeaturizer"
+          # Whether to add word vectors to ``ner_features``
+          ner_feature_vectors: False
```
So, if `ner_feature_vectors: False` then `ner_features` are set or not? `False` is kind of unintuitive. I would expect setting `ner_feature_vectors` to `True` would result in adding them.

tabergma replied:
And maybe also mention the default value.

jamesmf replied:
Oh I thought I had to put the default value here. It works the way you originally thought, but it's `False` by default. I'll update the docs to say `True` but note that it's not the default

tabergma commented on docs/nlu/entity-extraction.rst:
```diff
@@ -149,3 +149,12 @@ associate that with a location entity.
+If you want to pass custom features to ``CRFEntityExtractor``, you can create a ``Featurizer`` that provides ``ner_features``.
+If you do, ``ner_features`` should be an iterable of ``len(tokens)``, where each entry is a vector.
+If ``CRFEntityExtractor`` finds ``"ner_features"`` in one of the arrays in ``features`` in the config, it will pass the ``ner_features`` vectors to ``sklearn_crfsuite``.
+The simplest example of this is to pass word vectors as features, which you can do using ``SpacyFeaturizer``
```
```suggestion
The simplest example of this is to pass word vectors as features, which you can do using ``SpacyFeaturizer``.
```

tabergma replied:
Maybe also add a link here to the SpacyFeaturizer Component.

jamesmf commented:
Shoot I was busy this past week and it seems like meaningful conflicts may have arisen in that time period. 

It looks like the conflicts in the `/featurizers` files are both somewhat related, but I don't understand them well enough to properly resolve. Can I get some input?

amn41 commented:
hi @jamesmf ! @tabergma is on holiday this week so will be a few days before she can advise

dakshvar22 commented:
Hi @jamesmf ! I authored some of the changes(https://github.com/RasaHQ/rasa/pull/4233) because of which the merge conflicts might be arising, so I may be able to help you out here:

1. We replaced string hard-coding of feature names with constants which are defined in `rasa/nlu/constants.py`.
2. Feature computation of all attributes(text, intent, etc.) of a message object are front-loaded to corresponding component itself. Hence, you will see a refactored method - `_combine_with_existing_features()` with a new parameter `feature_name`. You can consider `ner_features` as a feature of the message, assign a constant to it and reuse that function.

jamesmf commented:
cool @dakshvar22 , thanks for responding! I took a shot at resolving the conflicts but am a bit confused on one piece. Does this new implementation imply that `SpacyFeaturizer` will `provide` every value in `MESSAGE_VECTOR_FEATURE_NAMES`? Do all `Featurizers` now provide all vectors? Is that desired? Or am I just misunderstanding the new `provides` line.

I'd like to add 
```
MESSAGE_ATTRIBUTES = [
    MESSAGE_TEXT_ATTRIBUTE,
    MESSAGE_INTENT_ATTRIBUTE,
    MESSAGE_RESPONSE_ATTRIBUTE,
    MESSAGE_ENTITIES_ATTRIBUTE,
]
...
MESSAGE_VECTOR_FEATURE_NAMES = {
    MESSAGE_TEXT_ATTRIBUTE: "text_features",
    MESSAGE_INTENT_ATTRIBUTE: "intent_features",
    MESSAGE_RESPONSE_ATTRIBUTE: "response_features",
    MESSAGE_ENTITIES_ATTRIBUTE: "ner_features",
}
```
but I assume that will affect other `Featurizers` outside of `SpacyFeaturizer`, which seems odd.

dakshvar22 commented:
Hi @jamesmf , 

> Do all Featurizers now provide all vectors

It's upto the featurizer to implement featurization of every attribute if it can. For example, `CountVectorFeaturizer` will not implement featurization of entities. In principle, all current featurizers can featurize text, intent and response attributes of a message.

I recommend adding `MESSAGE_ENTITIES_ATTRIBUTE` only to `MESSAGE_VECTOR_FEATURE_NAMES` since the `MESSAGE_ATTRIBUTES` is used universally across all featurizers to iterate over all attributes of message which they can featurize. Strictly entities is actually an attribute of the text part of a message.

jamesmf commented:
I'd love more guidance on this.

Theoretically there's nothing stopping every `Featurizer` that currently featurizes whole strings of text from provider `ner_features` by featurizing each `token`. This seems consistent with the "if it can vectorize text, use `[att for att in MESSAGE_ATTRIBUTES]`" kind of approach. But I'm guessing we don't want to add every text-based feature to the entity representation by default.

tabergma commented:
As @dakshvar22 suggested I would modify the constants like this:
```
...
MESSAGE_VECTOR_FEATURE_NAMES = {
    MESSAGE_TEXT_ATTRIBUTE: "text_features",
    MESSAGE_INTENT_ATTRIBUTE: "intent_features",
    MESSAGE_RESPONSE_ATTRIBUTE: "response_features",
    MESSAGE_ENTITIES_ATTRIBUTE: "ner_features",
}

MESSAGE_SPACY_FEATURES_NAMES = {
    MESSAGE_TEXT_ATTRIBUTE: "spacy_doc",
    MESSAGE_INTENT_ATTRIBUTE: "intent_spacy_doc",
    MESSAGE_RESPONSE_ATTRIBUTE: "response_spacy_doc",
    MESSAGE_ENTITIES_ATTRIBUTE: "spacy_doc",
}
...
```

So, we don't modify the `MESSAGE_ATTRIBUTES` list. Instead, I would add the attribute directly in the `SpacyFeaturizer`:

```
    provides = [
        MESSAGE_VECTOR_FEATURE_NAMES[attribute] for attribute in MESSAGE_ATTRIBUTES
    ] + MESSAGE_VECTOR_FEATURE_NAMES[MESSAGE_ENTITIES_ATTRIBUTE]

    requires = [
        MESSAGE_SPACY_FEATURES_NAMES[attribute] for attribute in MESSAGE_ATTRIBUTES
    ] + MESSAGE_SPACY_FEATURES_NAMES[MESSAGE_ENTITIES_ATTRIBUTE]
    
```

What do you think?

jamesmf commented:
Sure that works for me! I don't quite get the philosophy of it (all of the featurizers using `MESSAGE_ATTRIBUTES_FEATURE_NAMES` should be capable of providing entity features too) but it seems like a reasonable place to start to only have spacy provide `ner_features`.

tabergma commented on CHANGELOG.rst:
```diff
@@ -49,6 +49,10 @@ Added
 - Add a `bot_challenge` intent and a `utter_iamabot` action to all example projects and the rasa init bot.
 - Allow sending attachments when using the socketio channel
 - ``rasa data validate`` will fail with a non-zero exit code if validation fails
+- The tracker is now available within the interpreter's ``parse`` method, giving the ability to create interpreter classes that
```
Can be removed as it is duplicated. I guess, it is a leftover from a merge.

jamesmf replied:
Do I need to move this into the `1.4.0` section?

tabergma replied:
No, it is already part of section [1.3.0] - 2019-09-05. The entry is just duplicated, so need to be removed here.

tabergma commented on rasa/nlu/featurizers/__init__.py:
```diff
@@ -1,5 +1,7 @@
 import numpy as np
 
+from typing import Any
+from rasa.nlu.training_data import Message
```
Unused imports. Please remove them.

tabergma commented on rasa/nlu/featurizers/spacy_featurizer.py:
```diff
@@ -35,19 +36,28 @@ class SpacyFeaturizer(Featurizer):
 
     requires = [
         MESSAGE_SPACY_FEATURES_NAMES[attribute] for attribute in MESSAGE_ATTRIBUTES
     ]
```
```suggestion
    ] + [MESSAGE_SPACY_FEATURES_NAMES[MESSAGE_ENTITIES_ATTRIBUTE]]
```

tabergma commented:
Please update the branch and sign the Contributor License Agreement. Thanks!

jamesmf commented:
A lot of tests fail after I merged in master, as `spacy_utils` is trying to run `proprocess_text` on everything in `SPACY_FEATURIZABLE_ATTRIBUTES` but `entities` is obviously not a string. 

I can see that `entities` doesn't belong in  `SPACY_FEATURIZABLE_ATTRIBUTES` as it is implemented, but since it also defines the `requires` and `provides`, I don't know the best path forward. The last time I addressed that, the implementation changed.

For `_combine_with_existing_features` the implementation with `hstack` seems odd for single-token text. I need to look into that.

tabergma commented:
I would recommend to go back adding it manually in `SpacyFeaturizer` and remove `MESSAGE_ENTITIES_ATTRIBUTE` from `SPACY_FEATURIZABLE_ATTRIBUTES`. E.g.
```
    provides = [
        MESSAGE_VECTOR_FEATURE_NAMES[attribute]
        for attribute in SPACY_FEATURIZABLE_ATTRIBUTES
    ] + [MESSAGE_VECTOR_FEATURE_NAMES[MESSAGE_ENTITIES_ATTRIBUTE]]

    requires = [
        MESSAGE_SPACY_FEATURES_NAMES[attribute]
        for attribute in SPACY_FEATURIZABLE_ATTRIBUTES
    ] + [MESSAGE_SPACY_FEATURES_NAMES[MESSAGE_ENTITIES_ATTRIBUTE]]
```

jamesmf commented:
I think this is good to go

tabergma commented on CHANGELOG.rst:
```diff
@@ -112,6 +112,8 @@ Added
 - Add a `bot_challenge` intent and a `utter_iamabot` action to all example projects and the rasa init bot.
 - Allow sending attachments when using the socketio channel
 - ``rasa data validate`` will fail with a non-zero exit code if validation fails
+- `CRFEntityExtractor` updated to accept arbitrary token-level features like word vectors (issues/4214)
```
This ended up in the wrong section. Can you move it to the top of the changelog file into section 1.4.x? Thanks.

jamesmf commented:
@tabergma is feature gone with the sparse/dense + sequences refactor to featurizers? I haven't tested it but didn't see any evidence of it in the docs or in `CRFEntityExtractor`

tabergma commented:
@jamesmf The feature is still there, see https://rasa.com/docs/rasa/nlu/entity-extraction/#passing-custom-features-to-crfentityextractor. Any dense features will be taken into account as long as they are placed before the `CRFEntityExtractor` in the pipeline.

jamesmf commented:
oh that's really cool, great to see it generalized! thanks @tabergma 

I tried just adding `text_dense_features` to the `features` array, then `return_sequence` to the `SpacyFeaturizer` and it looks like it's working. Will the `return_sequence` make sure that the `CLS` token is still used for the intent classifiers?

tabergma commented:
Kind of, if the `CLS` token is present we take that, otherwise we take the mean of the sequential spacy feature vectors for the intent classifier (which is basically the same). We will remove the options `return_sequence` and `use_cls_token` in the next release again, to make it easier for the user.
